packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+
+```TS
+interface TableCellState<TCellData extends TableRecord> {
+  data: TCellData;
+}
+```
+
+This interface could possibly be expanded in the future to communicate relevant table state to the cell template, such as whether or not the row is selected.
+
+### `TableColumn<>`
+
+This abstract class is what a column web component (i.e. a slotted column) must extend.
+
+```TS
+abstract class TableColumn<TCellData extends TableRecord = TableRecord> {
+    // This method will produce the expected TableCellState that the cellTemplate expects as
+    // its source
+    abstract generateCellState(cellData: TCellData): TableCellState<TCellData>;
+
+    // The template to use to render the cell content for the column
+    abstract cellTemplate: ViewTemplate<TableCellState<TCellData>>;
+
+    // The style to apply to the cellTemplate
+    cellStyles?: ElementStyles;
+
+    // The keys that should be present in TCellData.
+    // This array is parallel with the keys returned from `getDataKeys()`.
+    readonly cellDataKeyNames: readonly string[];
+
+    // The keys from the row data that correlate to the data that will be in TCellData.
+    // This array is parallel with the keys specified by `dataKeyNames`.
+    abstract getRowDataKeys(): string[];
+
+    // Function that allows the table column to validate the type that gets created
+    // for the cell data. This should validate that the types in TCellData are correct
+    // for each key defined by `dataKeyNames`.
+    // Return `true` if the data is valid. Return `false` if the data is not valid.
+    abstract validateCellData(cellData: TCellData): boolean;
+}
+```
+
+_Note: The `TableColumn` class may be updated to support other features not covered in this HLD such as sorting and grouping._
+
+Given the above class, a series of column providers to handle basic use cases can be written within Nimble. For example, the `TableColumn` implementation we could create for rendering data as a read-only `NimbleTextField` could look like this:
+
+```TS
+type TableColumnTextCellData = StringColumnData<'value'>;
+
+// this interface is used to pass auxiliary configuration to access within the cellTemplate
+interface TableColumnTextCellState<TCellData extends TableRecord> extends TableCellState<TCellData> {
+    placeholder: string;
+}
+
+public class TableColumnText extends TableColumn<TableColumnTextCellData> {
+    ...
+
+    public generateCellState(cellData: TextColumnCellData): TableColumnTextCellState<TextColumnCellData> {
+        return { data: cellData, placeholder: this.placeholder };
+    }
+
+    public cellDataKeyNames = ['value'] as const;
+
+    @attr
+    public valueKey: string;
+
+    @attr
+    public placeholder: string; // Column auxiliary configuration
```

Comment:
I think we should observe all these attributes for changes and have a way to notify the parent table that their values have changed (and to re-render) i.e. fire a custom `column-configuration-change` event or something. The table can listen for those events generically and trigger a new render on raf, etc.

For example if the user changes the placeholder text attribute programmatically, the current cells will show the old placeholder value but if the user scrolls new cells will show the new placeholder.

We could do it like the wafer map by manually tracking it (plugins define placeholderChanged, etc and call a base class method which notifies) or we can brute force in the base class and just report on any observable property change: https://www.fast.design/docs/fast-element/observables-and-state/#external-observation

I think I lean towards the later so plugin authors can't accidentally forget to notify for attribute / observable property changes. But regardless of impl, maybe spec should discuss that as a requirement.

Reply:
Made a comment about how we would expect to handle notifications for things like the `placeholder` property, but made it clear that this spec would not be ironing out the details of that implementation/API.